sw/device/lib/dif/dif_padctrl.h

```diff
@@ -0,0 +1,157 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_PADCTRL_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_PADCTRL_H_
+
+#include <stdbool.h>
+
+// Generated.
+#include "padctrl_regs.h"
+
+#include "sw/device/lib/base/mmio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Pad attributes.
+ *
+ * Some attributes have defined meanings, others are implementation defined
+ * (`kDifPadCtrlAttrAdditionalFirst` and onwards). Not all pads will support all
+ * attributes.
+ */
+typedef enum dif_padctrl_attr {
+  kDifPadctrlAttrIOInvert = 0,
+  kDifPadctrlAttrOpenDrain,
+  kDifPadctrlAttrPullDown,
+  kDifPadctrlAttrPullUp,
+  kDifPadctrlAttrKeeper,
+  kDifPadctrlAttrWeakDrive,
+  kDifPadctrlAttrAdditionalFirst,
+  kDifPadctrlAttrLast = PADCTRL_PARAM_ATTRDW - 1,
+} dif_padctrl_attr_t;
+
+/**
+ * Generic padctrl DIF return codes
+ */
+typedef enum dif_padctrl_result {
+  kDifPadctrlOk = 0,    /**< Operation succeeded. */
+  kDifPadctrlError = 1, /**< Operation failed due to an unspecified error. It
+                           cannot be recovered or retried. */
+  kDifPadctrlBadArg = 2 /**< Operation failed due to a bad argument and can be
+                           retried. No register reads or writes have been
+                           performed */
+} dif_padctrl_result_t;
+
+/**
+ * Pad kinds.
+ *
+ * A pad can be an MIO or DIO pad
+ */
+typedef enum dif_padctrl_pad_kind {
+  kDifPadctrlPadMio = 0, /**< MIO (Multiplexed Input/Output) pad. Connected via
+                            pinmux. */
+  kDifPadctrlPadDio,     /**< DIO (Dedicated Input/Output) pad. Connected
+                            directly to peripheral inputs and outputs, bypassing
+                            pinmux. */
+} dif_padctrl_pad_kind_t;
+
+/**
+ * Pad index.
+ *
+ * Combined with `dif_padctrl_pad_kind_t` to refer to a specific pad.
+ */
+typedef uint32_t dif_padctrl_pad_index_t;
```

Comment:
Is there a particular reason this type is given a name, rather than using uint32_t directly? It doesn't give you any sort of type-safety, so I'm not sure that you gain much here.

Reply:
We did this in the PLIC, and I do think it helps to document what DIF parameters mean, even though it gives no extra type safety.

Reply:
Fair enough.

Reply:
Indeed I think it helps communicate intent (especially for things like local variables) even if it doesn't give you compiler enforcement of that intent.